File: 714.买卖股票的最佳时机含手续费.py
#
# @lc app=leetcode.cn id=714 lang=python3
#
# [714] 买卖股票的最佳时机含手续费
#

# @lc code=start
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        a, b = 0, -fee
        for i in range(1, len(prices)):
            a, b = max(a, a+prices[i]-prices[i-1]-fee, b+prices[i]-prices[i-1]), \
                   max(a-fee, b+prices[i]-prices[i-1])
        return max(a, b)
# @lc code=end

File: 714.买卖股票的最佳时机含手续费.cpp
/*
 * @lc app=leetcode.cn id=714 lang=cpp
 *
 * [714] 买卖股票的最佳时机含手续费
 */

// @lc code=start
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        int n = prices.size();
        int dp[n][2];
        dp[0][0] = -fee;
        dp[0][1] = 0;
        for (int i = 1; i < n; i++) {
            dp[i][0] = max(dp[i-1][0] + prices[i] - prices[i-1], dp[i-1][1] - fee);
            dp[i][1] = max(dp[i-1][0] + prices[i] - prices[i-1], dp[i-1][1]);
        }
        return dp[n-1][1];
    }
};
// @lc code=end

